opendevin/controller/agent_controller.py

```diff
@@ -274,12 +280,16 @@ async def step(self, i: int) -> bool:
 
         self.update_state_after_step()
 
-        # whether to await for user messages
         if isinstance(action, AgentTalkAction):
-            # await for the next user messages
-            user_message_observation = await self.wait_for_user_input()
-            logger.info(user_message_observation, extra={'msg_type': 'OBSERVATION'})
-            await self.add_history(action, user_message_observation)
+            await self.event_stream.add_event(action, EventSource.AGENT)
+            user_message = await self.wait_for_user_input()
+            logger.info(user_message, extra={'msg_type': 'ACTION'})
+            # FIXME: we're hacking a message action into a user message observation, for the benefit of CodeAct
+            await self.add_history(
+                action,
+                UserMessageObservation(user_message.content),
+                add_to_stream=False,
+            )
```

Comment:
So the user message is an action, which happens after a particular action... It's clearly a special case, though I'm not sure why it isn't more like a background observation, an obs of the environment. The user spoke: some part of the environment gave a signal, the agent notices it. But it looks like we've tried that one...

Reply:
I'm going to open an issue after this to simplify Message actions and observations. CodeAct is relying on a few things I want to refactor, so I need to talk to Xingyao. Currently we have:
* AgentTalkAction
* UserMessageObservation
* AgentMessageObservation
* MessageAction
* AgentEchoAction

Reply:
https://github.com/OpenDevin/OpenDevin/issues/1595